velox/functions/sparksql/DateTimeFunctions.h

```diff
@@ -372,4 +372,82 @@ struct AddMonthsFunction {
     result = daysSinceEpoch;
   }
 };
+
+template <typename T>
+struct NextDayFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& config,
+      const arg_type<Date>* /*startDate*/,
+      const arg_type<Varchar>* dayOfWeek) {
+    if (dayOfWeek != nullptr) {
+      try {
+        this->weekDay_ = getDayOfWeekFromString(
+            std::string_view(dayOfWeek->data(), dayOfWeek->size()));
+      } catch (const VeloxUserError&) {
+        invalidFormat_ = true;
+      }
+      isConstFormat_ = true;
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE int8_t
+  getDayOfWeekFromString(const std::string_view dayOfWeek) {
+    if (dayOfWeek.empty()) {
+      VELOX_USER_CHECK(!dayOfWeek.empty(), "Day of week should not be empty")
+    }
+    std::string input = {dayOfWeek.begin(), dayOfWeek.end()};
+    std::string lower_str = boost::algorithm::to_lower_copy(input);
+    if (weekMap_.contains(lower_str)) {
+      return weekMap_.at(lower_str);
+    } else {
+      VELOX_USER_FAIL("Illegal input for day of week: {}", dayOfWeek);
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE int64_t
+  getNextDateForDayOfWeek(int64_t startDay, int8_t dayWeek) {
+    return startDay + 1 + ((dayWeek - 1 - startDay) % 7 + 7) % 7;
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Date>& result,
+      const arg_type<Date>& startDate,
+      const arg_type<Varchar>& dayOfWeek) {
+    if (invalidFormat_) {
+      VELOX_USER_FAIL("Illegal input for day of week: {}", dayOfWeek);
+    }
+    if (!isConstFormat_) {
+      this->weekDay_ = getDayOfWeekFromString(
+          std::string_view(dayOfWeek.data(), dayOfWeek.size()));
+    }
+    auto dateTime = getDateTime(startDate);
+    int32_t year = getYear(dateTime);
+    int32_t month = getMonth(dateTime);
+    int32_t day = getDay(dateTime);
```

Comment:
year, month, day are only used for the error message? if so, don't compute these if there is no error

```
if (nextDay != (int32_t)nextDay) {
   ...
}
```